doc: improve NODE_NETWORK_LIMITED documentation per BIP159

jonatack opened the pull request:
Per BIP159, the definition is:

NODE_NETWORK_LIMITED || bit 10 (0x400) || If signaled, the peer <I>MUST</I> be capable of serving at least the last 288 blocks (~2 days).

Fix our documentation related to this, and add BIP159 mentions where relevant.

---

Originally this PR also tried to clarify the following, but it was dropped in response to review:

`NODE_NETWORK_LIMITED` is set by default, and it only signals a pruned or otherwise limited node if `NODE_NETWORK` is not also set. See the init setup logic or `IsLimitedPeer()` in `src/net_processing.cpp` for examples.

jonatack committed:
Improve NODE_NETWORK_LIMITED documentation per BIP159

Sjors commented:
ACK 39e15f6ca4fd44cc459d80b7a1540c03923906dd

It's good to clarify the description, because the term `NODE_NETWORK_LIMITED` has confused me before into assuming that a node that isn't limited doesn't use the flag. But instead it signals the minimum capability, and `NODE_NETWORK` is added when ready to signal full capability. Too late to rename them.

You can also see this in the `init.cpp`.

We start with just `NODE_NETWORK_LIMITED`:

https://github.com/bitcoin/bitcoin/blob/82ba50513425bf0568d4f9456282dc9713132490/src/init.cpp#L807-L811

And then add `NODE_NETWORK`, if possible:

https://github.com/bitcoin/bitcoin/blob/82ba50513425bf0568d4f9456282dc9713132490/src/init.cpp#L1733-L1739

We don't drop `NODE_NETWORK_LIMITED`.

kevkevinpal commented:
ACK [39e15f6](https://github.com/bitcoin/bitcoin/pull/31805/commits/39e15f6ca4fd44cc459d80b7a1540c03923906dd)

I agree with @maflcko maybe instead of using pruned node we can refer to the node as "network limited node"

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -704,7 +704,7 @@ class NetinfoRequestHandler : public BaseRequestHandler
         "           \"w\" - WITNESS: peer can be asked for blocks and transactions with witness data (SegWit)\n"
         "           \"c\" - COMPACT_FILTERS: peer can handle basic block filter requests (see BIPs 157 and 158)\n"
-        "           \"l\" - NETWORK_LIMITED: peer limited to serving only the last 288 blocks (~2 days)\n"
+        "           \"l\" - NETWORK_LIMITED: peer can serve at least the last 288 blocks (~2 days); may signal a pruned node if NETWORK is unset (see BIP 159)\n"
```
Thanks, updated.

jonatack commented:
Updated per review feedback (thanks!)

> It's good to clarify the description, because the term `NODE_NETWORK_LIMITED` has confused me before into assuming that a node that isn't limited doesn't use the flag. But instead it signals the minimum capability, and `NODE_NETWORK` is added when ready to signal full capability. Too late to rename them.

Well said -- same for me.

Sjors commented on src/bitcoin-cli.cpp:
```diff
@@ -704,7 +704,8 @@ class NetinfoRequestHandler : public BaseRequestHandler
         "           \"c\" - COMPACT_FILTERS: peer can handle basic block filter requests (see BIPs 157 and 158)\n"
-        "           \"l\" - NETWORK_LIMITED: peer limited to serving only the last 288 blocks (~2 days)\n"
+        "           \"l\" - NETWORK_LIMITED: peer can serve at least the last 288 blocks (~2 days); signals a pruned or otherwise\n"
+        "                                  limited peer (i.e., assumeutxo download in progress) if NETWORK is unset (see BIP 159)\n"
```
We should probably try to keep the output lines of `bitcoin-cli -netinfo help` below 80 characters. It doesn't wrap in a nice way.

jonatack replied:
Don't disagree but that would require reformatting most of the help.

jonatack replied:
For here kept it shorter than the longest line ("If that argument is passed...")

jonatack replied:
Shortened the line length in the last push.

Sjors replied:
Just the line you're touching is fine, yes. Thanks.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -704,7 +704,8 @@ class NetinfoRequestHandler : public BaseRequestHandler
         "           \"c\" - COMPACT_FILTERS: peer can handle basic block filter requests (see BIPs 157 and 158)\n"
-        "           \"l\" - NETWORK_LIMITED: peer limited to serving only the last 288 blocks (~2 days)\n"
+        "           \"l\" - NETWORK_LIMITED: peer can serve at least the last 288 blocks (~2 days); signals a pruned or otherwise\n"
+        "                                  limited peer (i.e., assumeutxo download in progress) if NETWORK is unset (see BIP 159)\n"
```
Thanks for the reminder, for some reason I keep thinking i.e. means "for example."

Are you suggesting s/i.e./e.g./, or removing the example? My thought was that this doc is a bit more user-facing, so more explanation or an example might be helpful.

jonatack replied:
Removed the example here.

Sjors replied:
"example given" is how I remember e.g.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -704,7 +704,8 @@ class NetinfoRequestHandler : public BaseRequestHandler
         "           \"c\" - COMPACT_FILTERS: peer can handle basic block filter requests (see BIPs 157 and 158)\n"
-        "           \"l\" - NETWORK_LIMITED: peer limited to serving only the last 288 blocks (~2 days)\n"
+        "           \"l\" - NETWORK_LIMITED: peer can serve at least the last 288 blocks (~2 days); signals a\n"
+        "                                  pruned or otherwise limited peer if NETWORK is unset (see BIP 159)\n"
```
Can you please clarify your comment -- what isn't correct?

ajtowns replied:
Dropping "; signals a pruned ... (see BIP 159)" seems clearer, and matches the definition of `NETWORK_LIMITED` provided by [BIP 159](https://github.com/bitcoin/bips/blob/master/bip-0159.mediawiki).

jonatack replied:
Done.

jonatack commented:
In response to review feedback, I've reduced the scope of this PR and updated the pull description. Originally this tried to clarify the following, but it is now dropped:

`NODE_NETWORK_LIMITED` is set by default, and it only signals a pruned or otherwise limited node if `NODE_NETWORK` is not also set (see the setup logic or `IsLimitedPeer()` in `src/net_processing.cpp` for an example).

jonasschnelli commented:
I haven't looked at this closely, but it could be that the implementation needs an update.
The important point is *fingerprinting resistance*.

During designing the BIP, we decided that LIMITED NODES should always cut off at HEAD-288 to avoid getting fingerprinted by leaking the prune depth.

See the BIP:

> Counter-measures for peer fingerprinting
> 
> Peers may have different prune depths (depending on the peers configuration, disk space, etc.) which can result in a fingerprinting weakness (finding the prune depth through getdata requests). NODE_NETWORK_LIMITED supporting peers SHOULD avoid leaking the prune depth and therefore not serve blocks deeper than the signaled NODE_NETWORK_LIMITED threshold (288 blocks).
```

jonatack commented:
@jonasschnelli Good point. Yes. A BIP159 node MUST be able to serve at least the latest 288 blocks, but SHOULD not serve more. Will update.

jonatack commented:
> During designing the BIP, we decided that LIMITED NODES should always cut off at HEAD-288 to avoid getting fingerprinted by leaking the prune depth.

Maybe the BIP and documentation here should clarify that limited nodes are defined by both the limited bit being set and by the node network bit not being set, as described in https://github.com/bitcoin/bitcoin/pull/31805#issuecomment-2654830209.

jonatack commented:
Opened https://github.com/bitcoin/bips/pull/1768 and putting this in draft for now.

fanquake commented:
> Opened https://github.com/bitcoin/bips/pull/1768 and putting this in draft for now.

That has been merged.

fanquake commented:
What is the status of this?

jonatack commented:
Will push a wider update based on the BIP159 since https://github.com/bitcoin/bips/pull/1768.
